lib/src/boollist.dart

```diff
@@ -0,0 +1,252 @@
+  }
+
+  @override
+  int get length => _length;
+
+  @override
+  bool operator [](int index) {
+    RangeError.checkValidIndex(index, this, 'index', _length);
+    return (_data[index >> _entryShift] &
+            (1 << (index & _entrySignBitIndex))) !=
+        0;
+  }
+
+  @override
+  void operator []=(int index, bool val) {
+    RangeError.checkValidIndex(index, this, 'index', _length);
+    _setBit(index, val);
+  }
+
+  @override
+  void fillRange(int start, int end, [bool fill]) {
+    RangeError.checkValidRange(start, end, _length);
+    fill ??= false;
+
+    var startWord = start >> _entryShift;
+    var endWord = (end - 1) >> _entryShift;
+
+    var startBit = start & _entrySignBitIndex;
+    var endBit = (end - 1) & _entrySignBitIndex;
+
+    if (startWord < endWord) {
+      if (fill) {
+        _data[startWord] |= -1 << startBit;
+        _data.fillRange(startWord + 1, endWord, -1);
+        _data[endWord] |= (1 << (endBit + 1)) - 1;
+      } else {
+        _data[startWord] &= (1 << startBit) - 1;
+        _data.fillRange(startWord + 1, endWord, 0);
+        _data[endWord] &= -1 << (endBit + 1);
+      }
+    } else {
+      if (fill) {
+        _data[startWord] |= ((1 << (endBit - startBit + 1)) - 1) << startBit;
+      } else {
+        _data[startWord] &= ((1 << startBit) - 1) | (-1 << (endBit + 1));
+      }
+    }
+  }
+
+  /// Returns custom iterator for [BoolList].
+  ///
+  /// To provide null safety [Iterator.current] getter of returned iterator
+  /// returns `false` before and after iteration process.
+  @override
+  Iterator<bool> get iterator => _BoolListIterator(this);
+
+  void _setBit(int index, bool val) {
+    if (val) {
+      _data[index >> _entryShift] |= 1 << (index & _entrySignBitIndex);
+    } else {
+      _data[index >> _entryShift] &= ~(1 << (index & _entrySignBitIndex));
+    }
+  }
+
+  static int _lengthInWords(int bitsLength) {
+    return (bitsLength + (_bitsPerEntry - 1)) >> _entryShift;
+  }
+}
+
+class _GrowableBoolList extends BoolList {
+  static const int _growthFactor = 2;
+
+  _GrowableBoolList(int length)
+      : super._(
+          Uint32List(BoolList._lengthInWords(length * _growthFactor)),
+          length,
+        );
+
+  @override
+  set length(int length) {
+    RangeError.checkNotNegative(length, 'length');
+    if (length > _length) {
+      _expand(length);
+    } else if (length < _length) {
+      _shrink(length);
+    }
+  }
+
+  void _expand(int length) {
+    if (length > _data.length * BoolList._bitsPerEntry) {
+      _data = Uint32List(
+        BoolList._lengthInWords(length * _growthFactor),
+      )..setAll(0, _data);
+    }
+    _length = length;
+  }
+
+  void _shrink(int length) {
+    if (length < _length ~/ _growthFactor) {
+      var newDataLength = BoolList._lengthInWords(length);
+      _data = Uint32List(newDataLength)..setRange(0, newDataLength, _data);
+    }
+
+    for (var i = length; i < _data.length * BoolList._bitsPerEntry; i++) {
```

Comment:
Or
```dart
   _data[_data.length - 1] &= (1 << (length & (_bitsPerEntry - 1))) - 1;
```
to clear the remaining bits of the last word.

Reply:
Unfortunately:
1. This will failed in case where:
- new `length` is 0 (`_data[-1]`);
- `length` equals `_bitsPerEntry * _data.length`, last value in `_data` will be set to 0.
2. Growing/shrinking algorithm not replaces `_data` when new `length` is not less than `_length ~/ _growthFactor`. In this case cleaning bits in the last value of `_data` wrong, we need to clean all bits from the 'length' till last bit.

Reply:
ACK. Edge cases. Doing too much checking ahead of time to avoid the edge cases might end up taking more time than just doing the brute-force thing.